---
title: "Report: `r params$project`"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
    toc_depth: 3
params:
  project: test-gwas
  date: 2021-08-09
  version: v0.4.2
  regenie_merged: ../output/test-gwas/results/Y1.regenie.gz
  regenie_filename: test-gwas.Y1.regenie.all.gz
  phenotype: Y1
  covariates:
  condition_list:
  interaction_gxe: 
  interaction_gxg:
  phenotype_file: ../output/test-gwas/validated_input/phenotype.validated.txt
  regenie_step1_log: ../output/test-gwas/logs/test-gwas-additive.step1.log
  regenie_step2_log: ../output/test-gwas/logs/test-gwas-additive.step2.log
  phenotype_log: ../output/test-gwas/logs/phenotype.validated.log
  covariate_log: NO_COV_LOG
  plot_ylimit: 0
  manhattan_annotation_enabled: true
  annotation_min_log10p: 5
  annotated_tophits_filename: ../output/test-gwas/results/tophits/Y1.regenie.filtered.annotated.txt.gz
  r_functions: ./functions.R
  rmd_pheno_stats: child_phenostatistics.Rmd
  rmd_valdiation_logs: child_validationlogs.Rmd
---

```{r setup, include=FALSE}
library(DT)
library(data.table)
library(R.utils)
library(ggplot2)
library(skimr)
library(kableExtra)
library(dplyr)
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, error = TRUE, cache=FALSE)
source(params$r_functions)
```

## Project Summary

| Parameter        | Value                       |
|------------------|-----------------------------|
| Project          | `r params$project`          |
| Pipeline Version | `r params$version`          |
| Date             | `r params$date`             |
| Phenotype File   | `r params$phenotype_file`   |
| Phenotype        | `r params$phenotype`        |
| Covariates       | `r params$covariates`       |
| Condition list   | `r params$condition_list`   |
| Regenie Output   | `r params$regenie_filename` |

## Phenotype Statistics

```{r child=params$rmd_pheno_stats}

```

## Manhattan Plot

```{r, dpi=300}
regenieTable <- fread(params$regenie_merged, select = c("CHROM","GENPOS","LOG10P","ID", "A1FREQ"), header=TRUE, tmpdir="$PWD")
annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("ID", "GENE_NAME"), header=TRUE, tmpdir="$PWD")

#merge gene names of tophits to regenieTable
annotatedTopHits <- annotatedTopHits %>%  
    mutate(ID = as.character(ID))
regenieTable <- regenieTable %>%  
    mutate(ID = as.character(ID))

regenieTable_annotated <- left_join(regenieTable, annotatedTopHits, by = "ID")

#prepare data for plotting
df_prep <- manhattan_prep(regenieTable_annotated)

rm(regenieTable_annotated)

#create plot
p <- manhattan_plot(df_prep,
                    ymax = params$plot_ylimit)

#note for manual ylim setting (if ymax is defined)
if(params$plot_ylimit == 0) {
  limit = NULL
} else {
  limit = c(0, params$plot_ylimit )
    cat(paste("Please note that the Y-axis limit has been set to '", params$plot_ylimit,"'. Please use the 'plot_ylimit' parameter to adapt this behaviour."))
}

#annotatate plot (if it is enabled)
 if(!params$manhattan_annotation_enabled) {
  p
} else {
  p + ggrepel::geom_text_repel(data = df_prep %>% group_by(GENE_NAME) %>%
                             top_n(1, LOG10P) %>%
                               slice(1) %>%
                             filter(LOG10P > params$annotation_min_log10p),
                            aes(label=GENE_NAME), color = 1, fontface = "italic")
}
```

## QQ Plot

```{r, results='asis', dpi=300}
if(params$plot_ylimit == 0) {
  limit = NULL
} else {
  limit = c(0, params$plot_ylimit )
  pp <- pp + ylim(0, params$plot_ylimit)
  cat(paste("Please note that the Y-axis limit has been set to '", params$plot_ylimit,"'. Please use the 'plot_ylimit' parameter to adapt this behaviour."))
}

cat('\n\n')
cat('### Overall')
cat('\n\n')
pp <- qq_plot(df=regenieTable)
print(pp)

cat('\n\n')
cat('### Split by MAF')
cat('\n\n')
pp <- qq_plot(df=regenieTable,
              mafsplitting = TRUE)
print(pp)

```

Please note that the QQ Plot is split by the following MAFs if any variants fall in the respective interval:

* (0, 0.0005]
* (0.0005, 0.005]
* (0.005, 0.05]
* (0.05, 1]

## Top Loci

Filtered for a minimun -log<sub>10</sub>(p-value) = `r params$annotation_min_log10p`.

```{r}

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("CHROM", "GENPOS", "LOG10P", "ID", "GENE_NAME", "GENE_CHROMOSOME", "GENE_START", "GENE_END", "DISTANCE"), header=TRUE, tmpdir="$PWD")

topHitsGroupedByGene <-
  annotatedTopHits %>%
  group_by(GENE_NAME, GENE_CHROMOSOME, GENE_START, GENE_END) %>%
  summarise(GENE_LOG10P = max(LOG10P), VARIANTS = n()) %>%
  arrange(desc(GENE_LOG10P))

datatable(topHitsGroupedByGene)

```

## Validation and Logs

```{r child=params$rmd_valdiation_logs}

```


---

<small>
This report has been created with **[nf-gwas `r params$version`](https://github.com/genepi/nf-gwas)**, a nextflow pipeline developed by [Sebastian Sch√∂nherr](mailto:sebastian.schoenherr@i-med.ac.at), [Lukas Forer](mailto:lukas.forer@i-med.ac.at) and [Johanna Schachtl-Riess](mailto:johanna.schachtl-riess@i-med.ac.at), Institute of Genetic Epidemiology, Medical University of Innsbruck.
</small>
