---
title: "`r params$phenotype` (`r params$project`)"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
    toc_depth: 3
params:
  project: test-gwas
  date: 2021-08-09
  version: v0.4.2
  regenie_merged: ../output/test-gwas/results/Y1.regenie.gz
  regenie_filename: test-gwas.Y1.regenie.all.gz
  phenotype: Y1
  covariates:
  condition_list:
  interaction_gxe:
  interaction_gxg:
  phenotype_file: ../output/test-gwas/validated_input/phenotype.validated.txt
  regenie_step1_log: ../output/test-gwas/logs/test-gwas-additive.step1.log
  regenie_step2_log: ../output/test-gwas/logs/test-gwas-additive.step2.log
  phenotype_log: ../output/test-gwas/logs/phenotype.validated.log
  covariate_log: NO_COV_LOG
  plot_ylimit: 0
  manhattan_annotation_enabled: true
  annotation_min_log10p: 5
  annotated_tophits_filename: ../output/test-gwas/results/tophits/Y1.regenie.filtered.annotated.txt.gz
  binned_results:
  r_functions: ./functions.R
  rmd_pheno_stats: child_phenostatistics.Rmd
  rmd_valdiation_logs: child_validationlogs.Rmd
---

```{r setup, include=FALSE}
library(DT)
library(data.table)
library(R.utils)
library(ggplot2)
library(skimr)
library(kableExtra)
library(dplyr)
library(tidyr)
library(ramwas)
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, error = TRUE, cache=FALSE)
source(params$r_functions)
```

## Project Summary

| Parameter        | Value                       |
|------------------|-----------------------------|
| Project          | `r params$project`          |
| Pipeline Version | `r params$version`          |
| Date             | `r params$date`             |
| Phenotype File   | `r params$phenotype_file`   |
| Phenotype        | `r params$phenotype`        |
| Covariates       | `r params$covariates`       |
| Condition list   | `r params$condition_list`   |
| Regenie Output   | `r params$regenie_filename` |

## Phenotype Statistics

```{r child=params$rmd_pheno_stats}

```


## Manhattan Plot

```{r, dpi=300}
resultsBinned <- fread(params$binned_results, header=TRUE, tmpdir="temp")

resultsBinned$LOG10P <- case_when(!is.na(resultsBinned$LOG10P) == TRUE ~ resultsBinned$LOG10P,
                                    TRUE ~ resultsBinned$y1)

#prepare data for plotting
df_prep <- manhattan_prep(resultsBinned)

#create plot
p <- manhattan_plot(df_prep,
                    ymax = params$plot_ylimit)

#note for manual ylim setting (if ymax is defined)
if(params$plot_ylimit == 0) {
  limit = NULL
} else {
  limit = c(0, params$plot_ylimit )
    cat(paste("Please note that the Y-axis limit has been set to '", params$plot_ylimit,"'. Please use the 'plot_ylimit' parameter to adapt this behaviour."))
}

#annotatate plot (if it is enabled)
 if(!params$manhattan_annotation_enabled) {
  p
} else {
  p + ggrepel::geom_text_repel(data = df_prep %>% filter(type == "peak"),
                            aes(label=GENE_NAME), color = 1, fontface = "italic")
}
```

## Significant loci

Loci are defined ±200 kb around lead SNPs. Significance level is set to -log<sub>10</sub>(p-value) = `r params$annotation_min_log10p`.

```{r}
results <- fread(params$regenie_merged, header=TRUE, tmpdir="temp")

results <- results %>%
  unite("Position", 
        CHROM:GENPOS, 
        sep=":",
        remove = FALSE)

resultsBinned <- resultsBinned %>%
  unite("Position", 
        CHROM:GENPOS, 
        sep=":",
        remove = FALSE)

# merge peak info to all significant SNPs
results_peak <- left_join(results %>% filter(LOG10P > params$annotation_min_log10p), 
                          resultsBinned %>% select(Position, type) %>% filter(type == "peak"),
                          by = "Position")

# create peak range +- 200kb
peak_range <- resultsBinned %>% 
  select(CHROM, GENPOS, Position, type) %>% 
  filter(type == "peak") %>% 
  rename(peak_group = Position)
peak_range$min <- peak_range$GENPOS - 200000
peak_range$max <- peak_range$GENPOS + 200000
peak_range <- peak_range %>%
  select(CHROM, peak_group, min, max)

#assign peak_group to significant SNPs
peak_range <- setDT(peak_range)
results_peak <- setDT(results_peak)
results_peakgroup <- peak_range[results_peak, on = .(CHROM, min <= GENPOS, max >= GENPOS)]
results_peakgroup$GENPOS <- results_peakgroup$min

rm(results_peak)
rm(peak_range)

#calculate no. of significant variants pear peak_group
results_peakgroup <- results_peakgroup %>%
  group_by(peak_group)

topHitsN <- 
  results_peakgroup %>%
  summarise(VARIANTS = n())

#filter out peak/lead variants
topHitsLead <- results_peakgroup %>%
  filter(type == "peak")


#join variant no. to lead variants and restructure
lociByPeak <- right_join(topHitsN, topHitsLead,
                         by = c("peak_group" = "Position")) %>%
  ungroup() %>%
  select(GENE_NAME, VARIANTS, CHROM, GENPOS, RSID, ALLELE0, ALLELE1, A1FREQ, BETA, SE, LOG10P, GENE_DISTANCE, N, TEST) %>%
  unite("Position", CHROM:GENPOS, sep=":")  %>%
  unite("Allele 0/1", ALLELE0:ALLELE1, sep="/")  %>%
  arrange(desc(LOG10P))

# create custom table container
sketch = htmltools::withTags(table(
  class = 'display',
  thead(
    tr(
      th(rowspan = 2, ''),
      th(rowspan = 2, 'Nearest gene'),
      th(rowspan = 2, 'No. of signif. SNPs'),
      th(colspan = 11, 'lead SNP')
      ),
    tr(
      th('Position'),
      th('rsID'),
      th('Allele 0/1'),
      th('Allele 1 Freq'),
      th('Beta'),
      th('Standard error (SE)'),
      th('-log10(p)'),
      th('Distance to gene'),
      th('N'),
      th('Test')
    )
  )
))

# create table
datatable(lociByPeak,
          container = sketch,
          rownames = TRUE,
          width = "100%")

# output number of toploci for phenotype
loci <- nrow(lociByPeak)
toplociN <- data.frame(params$phenotype, loci)
write.table(toplociN, 
            quote = FALSE,
            row.names = FALSE,
            file = paste("loci_", params$regenie_filename, ".txt",
                       sep = ""))

```

## QQ Plot

```{r, results='asis', dpi=300}

qqPrepare <- qqPlotPrepare(pvalues=results$LOG10P,ismlog10 = T)
qqPlotFast(qqPrepare)

```

## Validation and Logs

```{r child=params$rmd_valdiation_logs}

```


---

<small>
This report has been created with **[nf-gwas `r params$version`](https://github.com/genepi/nf-gwas)**, a nextflow pipeline developed by [Sebastian Schönherr](mailto:sebastian.schoenherr@i-med.ac.at), [Lukas Forer](mailto:lukas.forer@i-med.ac.at), [Johanna Schachtl-Riess](mailto:johanna.schachtl-riess@i-med.ac.at) and [Silvia Di Maio](mailto:silvia.di-maio@i-med.ac.at), Institute of Genetic Epidemiology, Medical University of Innsbruck.
</small>
